src/driver/src/WorkloadParser.cpp

```diff
@@ -0,0 +1,190 @@
+
+    if (_isSmokeTest) {
+        // Do a second pass to convert config into the smoke test version.
+        parsedWorkload = recursiveParse(parsedWorkload, ParseMode::kSmokeTest);
+    }
+
+    return parsedWorkload;
+}
+
+YAML::Node WorkloadParser::loadFile(const std::string& source) {
+    try {
+        return YAML::LoadFile(source);
+    } catch (const std::exception& ex) {
+        BOOST_LOG_TRIVIAL(error) << "Error loading yaml from " << source << ": " << ex.what();
+        throw;
+    }
+}
+
+YAML::Node WorkloadParser::recursiveParse(YAML::Node node, ParseMode mode) {
+    YAML::Node out;
+    switch (node.Type()) {
+        case YAML::NodeType::Map: {
+            for (auto kvp : node) {
+                if (mode == ParseMode::kSmokeTest)
+                    convertToSmokeTest(kvp.first.as<std::string>(), kvp.second, out);
+                else if (mode == ParseMode::kNormal)
+                    convertExternal(kvp.first.as<std::string>(), kvp.second, out);
+            }
+            break;
+        }
+        case YAML::NodeType::Sequence: {
+            for (auto val : node) {
+                out.push_back(recursiveParse(val, mode));
+            }
+            break;
+        }
+        default:
+            return node;
+    }
+    return out;
+}
+
+YAML::Node WorkloadParser::replaceParam(YAML::Node input) {
+    if (!input["Name"] || !input["Default"]) {
+        auto os = std::ostringstream();
+        os << "Invalid keys for '^Parameter', please set 'Name' and 'Default' in following node"
+           << YAML::Dump(input);
+        throw InvalidConfigurationException(os.str());
+    }
+
+    auto name = input["Name"].as<std::string>();
+    // The default value is mandatory.
+    auto defaultVal = input["Default"];
+
+    // Nested params are ignored for simplicity.
+    if (auto paramVal = _params.find(name); paramVal != _params.end()) {
+        return paramVal->second;
+    } else {
+        return input["Default"];
+    }
+}
+
+void WorkloadParser::convertExternal(std::string key, YAML::Node value, YAML::Node& out) {
+    if (key == "^Parameter") {
+        out = replaceParam(value);
+    } else if (key == "ExternalPhaseConfig") {
+        auto external = parseExternal(value);
+        // Merge the external node with the any other parameters specified
+        // for this node like "Repeat" or "Duration".
+        for (auto externalKvp : external) {
+            if (!out[externalKvp.first])
+                out[externalKvp.first] = externalKvp.second;
+        }
+    } else {
+        out[key] = recursiveParse(value, ParseMode::kNormal);
+    }
+}
+
+void WorkloadParser::convertToSmokeTest(std::string key, YAML::Node value, YAML::Node& out) {
+    if (key == "Duration" || key == "Repeat") {
+        out["Repeat"] = 1;
+    } else if (key == "Rate" || key == "SleepBefore" || key == "SleepAfter") {
+        // Ignore those keys in smoke tests.
+    } else {
+        out[key] = recursiveParse(value, ParseMode::kSmokeTest);
```

Comment:
Since this applies at all depths of the traversal, would this prevent something like this?

```yaml
Actors:
- Type: MyActor
  Phases:
  - Document: { Duration: 7 }
```

↑ is contrived but not unreasonable...I think the `convertToSmokeTest` may need to only be applied to top-level, `Actor`, and `Phase` blocks or something.

It also doesn't seem unreasonable to require these "reserved" keys to be placed inside `{^Verbatim}` blocks and `recursiveParse` could change the `ParseMode` if it sees that. The syntax would be funny:

```yaml
Actors:
- Type: MyActor
  Phases:
  - Document: {^Verbatim: { Duration: 7 }}
  # or maybe this?
  - Document: {Duration: {^Verbatim: 7}}
```

but we'd also want a way to document or enforce that somehow I think.

(Also tbh this isn't a *huge* deal since dry-run is really just to catch bizarro mis-configurations and the likelihood of an Actor using these "keywords" in its own configuration is admittedly low.)